src/cli/commands/publish.js

```diff
@@ -118,6 +119,22 @@ async function publish(config: Config, pkg: any, flags: Object, dir: string): Pr
 }
 
 export async function run(config: Config, reporter: Reporter, flags: Object, args: Array<string>): Promise<void> {
+  // validate arguments
+  const dir = args[0] ? path.resolve(config.cwd, args[0]) : config.cwd;
+  if (args.length > 1) {
+    throw new MessageError(reporter.lang('tooManyArguments', 1));
+  }
+  if (!await fs.exists(dir)) {
+    throw new MessageError(reporter.lang('unknownFolderOrTarball'));
+  }
+
+  const stat = await fs.lstat(dir);
+  let publishPath = dir;
+  if (stat.isDirectory()) {
+    config.cwd = path.resolve(dir);
```

Comment:
I'm not fond of changing the `cwd` outside of the entry point, but since `publish` is a very specific feature ... Ideally I'd prefer if we were passing the publish path around.